clang/test/CIR/CodeGen/builtins-memory.c

```diff
@@ -0,0 +1,56 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir %s -o %t.cir
+// RUN: FileCheck %s --check-prefix=CIR --input-file=%t.cir
+
+typedef __SIZE_TYPE__ size_t;
+void test_memcpy_chk(void *dest, const void *src, size_t n) {
+  // CIR-LABEL: cir.func @test_memcpy_chk
+  // CIR:         %[[#DEST:]] = cir.alloca {{.*}} ["dest", init]
+  // CIR:         %[[#SRC:]] = cir.alloca {{.*}} ["src", init]
+  // CIR:         %[[#N:]] = cir.alloca {{.*}} ["n", init]
+
+  // memcpy should be emitted when the count is known to fit the buffer size.
+
+  // CIR: %[[#DEST_LOAD:]] = cir.load %[[#DEST]]
+  // CIR: %[[#SRC_LOAD:]] = cir.load %[[#SRC]]
+  // CIR: %[[#COUNT:]] = cir.const #cir.int<8>
+  // CIR: cir.libc.memcpy %[[#COUNT]] bytes from %[[#SRC_LOAD]] to %[[#DEST_LOAD]]
+  __builtin___memcpy_chk(dest, src, 8, 10);
+
+  // CIR: %[[#DEST_LOAD:]] = cir.load %[[#DEST]]
+  // CIR: %[[#SRC_LOAD:]] = cir.load %[[#SRC]]
+  // CIR: %[[#COUNT:]] = cir.const #cir.int<10>
+  // CIR: cir.libc.memcpy %[[#COUNT]] bytes from %[[#SRC_LOAD]] to %[[#DEST_LOAD]]
+  __builtin___memcpy_chk(dest, src, 10, 10);
+
+  // A count known to be smaller than the size should call the libc function.
+
+  // CIR: %[[#DEST_LOAD:]] = cir.load %[[#DEST]]
+  // CIR: %[[#SRC_LOAD:]] = cir.load %[[#SRC]]
+  // CIR: %[[#COUNT:]] = cir.const #cir.int<10>
+  // CIR: %[[#SIZE:]] = cir.const #cir.int<8>
+  // CIR: cir.call @__memcpy_chk(%[[#DEST_LOAD]], %[[#SRC_LOAD]], %[[#COUNT]], %[[#SIZE]])
```

Comment:
I'd prefer if we model this with an extra optional operand in cir.libc.memcpy and just lower that to the proper call during LLVM lowering. This might allow us later to fold the check into the most regular `cir.libc.memcpy` form if profitable and safe.

Reply:
And because of this, you might as well to just check all LLVM outputs in this file as well.

Reply:
What we discussed offline: this is good enough and we can make additional improvements later